### PR TITLE
feat: ml pipelines

### PR DESCRIPTION
**Pull request by nkoppisetty**
Explain what this PR does.

Adding ml_pipelines and a pipeline udf for configuring multiple ML pipelines for a single stream config.

### CODE REVIEW

File: numalogic/udfs/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    def compute(cls, model: artifact_t, input_: npt.NDArray[float], **kwargs):
+        pass
+
+    def __init__(self, pl_conf: Optional[PipelineConf] = None):
+        super().__init__(pl_conf=pl_conf, _vtx="pipeline")
+
+    @UDF_TIME.time()
+    def exec(self, keys: list[str], datum: Datum) -> Messages:
+        """
+        The preprocess function here receives data from the data source.
```
Comment (s0nicboOm) [inline numalogic/udfs/pipeline.py:42]: knitpick: can we update the docstring please.

File: numalogic/udfs/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+from numalogic.tools.types import artifact_t
+from numalogic.udfs._metrics import UDF_TIME
+from numalogic.udfs import NumalogicUDF
+from numalogic.udfs._config import PipelineConf
+
+# TODO: move to config
+LOCAL_CACHE_TTL = int(os.getenv("LOCAL_CACHE_TTL", "3600"))
+LOCAL_CACHE_SIZE = int(os.getenv("LOCAL_CACHE_SIZE", "10000"))
+LOAD_LATEST = os.getenv("LOAD_LATEST", "false").lower() == "true"
```
Comment (s0nicboOm) [inline numalogic/udfs/pipeline.py:19]: Not needed.

File: numalogic/udfs/preprocess.py [hunk 3]
```diff
@@ -95,17 +95,16 @@ def exec(self, keys: list[str], datum: Datum) -> Messages:
...
             return Messages(Message.to_drop())
-        raw_df, timestamps = get_df(
-            data_payload=data_payload, stream_conf=self.get_conf(data_payload["config_id"])
-        )
+
+        _stream_conf = self.get_stream_conf(data_payload["config_id"])
+        _conf = _stream_conf.ml_pipelines[data_payload["pipeline_id"]]
+        raw_df, timestamps = get_df(data_payload=data_payload, stream_conf=_stream_conf)
 
         _metric_label_values = (self._vtx, ":".join(keys), data_payload["config_id"])
```
Comment (s0nicboOm) [inline numalogic/udfs/preprocess.py:103]: I think for metrics we need to have pipeline_id also an label. Can you please add a TODO and open an issue and assign it to me.

**Review by s0nicboOm — CHANGES_REQUESTED** [review]
minor comments

File: numalogic/udfs/pipeline.py
[the remark this answers is not in the document]
Comment (nkoppisetty) [reply numalogic/udfs/pipeline.py:1]: yeah couldn't think of anything, let me know if you have something on mind

[the remark this answers is not in the document]
Comment (nkoppisetty) [reply numalogic/udfs/pipeline.py:1]: named it ml_pipeline for now, feel free to update it later with whatever suits best

**Comment by s0nicboOm** [comment]
Merged the changes after refactoring: https://github.com/numaproj/numalogic/pull/336

**Pull request closed without merging**